### PR TITLE
product-details: don't ask hg.mozilla.org for data we already have (bug 1936479)

### LINKED ISSUE #1582: Product-details did not update with 133.0.1
>Hi relduty, https://product-details.mozilla.org/1.0/mobile_versions.json wasn't updated with 133.0.1, I am not seeing any failure in the ship_android graph above. I am happy to click on the "update product-details" button in Ship-it but before I do that, do you want to look at it in case rebuilding would erase some log information?

### PR DESCRIPTION
**Pull request by jcristau**
The l10n data for a shipped release is immutable, so there's no point gathering it over and over.

This means get_primary_builds() needs to look at old_product_details as well.

Fixes #1582

### CODE REVIEW

**Comment by jcristau** [comment]
Output from local testing:
```
$ docker exec -ti shipit_api_1 /bin/bash
app@afd2542b9914:~$ DEBUG=1 poetry run shipit_rebuild_product_details --database-url="postgresql://shipituser:shipitpassword@db/shipitdb" --channel production
Product details are building ...
2024-12-11 13:42:56,611 - INFO: Getting latest product details from https://github.com/mozilla-releng/product-details.
2024-12-11 13:42:56,611 - DEBUG: Running command: git clone -b production https://github.com/mozilla-releng/product-details product-details, {'stdin': -3, 'stdout': -1, 'stderr': -1, 'cwd': PosixPath('/tmp')}
2024-12-11 13:43:01,965 - DEBUG: Running command: git config http.postBuffer 12M, {'stdin': -3, 'stdout': -1, 'stderr': -1, 'cwd': PosixPath('/tmp/product-details')}
2024-12-11 13:43:01,971 - DEBUG: Running command: git config user.email release-services+robot@mozilla.com, {'stdin': -3, 'stdout': -1, 'stderr': -1, 'cwd': PosixPath('/tmp/product-details')}
2024-12-11 13:43:01,980 - DEBUG: Running command: git config user.name Release Services Robot, {'stdin': -3, 'stdout': -1, 'stderr': -1, 'cwd': PosixPath('/tmp/product-details')}
2024-12-11 13:43:01,984 - INFO: Breakpoint version is 114
2024-12-11 13:43:01,985 - INFO: Reading old product details from /tmp/product-details/public
2024-12-11 13:43:02,246 - INFO: Getting old releases from the database
2024-12-11 13:43:02,286 - INFO: Getting the current nightly version from the database
2024-12-11 13:43:02,289 - INFO: Getting locales from hg.mozilla.org for each release from database
2024-12-11 13:43:02,295 - DEBUG: Fetching https://hg.mozilla.org/releases/mozilla-beta/raw-file/9fb87e89c26069198ce2a59a0a790a264d225169/browser/locales/l10n-changesets.json
2024-12-11 13:43:02,295 - DEBUG: Fetching https://hg.mozilla.org/releases/mozilla-release/raw-file/7ed49fe90e846a6d836cf434ba175068e977dc8e/browser/locales/l10n-changesets.json
2024-12-11 13:43:04,071 - DEBUG: Fetched https://hg.mozilla.org/releases/mozilla-release/raw-file/7ed49fe90e846a6d836cf434ba175068e977dc8e/browser/locales/l10n-changesets.json
2024-12-11 13:43:05,377 - DEBUG: Fetched https://hg.mozilla.org/releases/mozilla-beta/raw-file/9fb87e89c26069198ce2a59a0a790a264d225169/browser/locales/l10n-changesets.json
2024-12-11 13:43:05,452 - DEBUG: Fetching https://hg.mozilla.org/mozilla-central/raw-file/default/browser/locales/l10n-changesets.json
2024-12-11 13:43:05,452 - DEBUG: Fetching https://hg.mozilla.org/comm-central/raw-file/default/mail/locales/l10n-changesets.json
2024-12-11 13:43:05,703 - DEBUG: Fetched https://hg.mozilla.org/comm-central/raw-file/default/mail/locales/l10n-changesets.json
2024-12-11 13:43:07,525 - DEBUG: Fetched https://hg.mozilla.org/mozilla-central/raw-file/default/browser/locales/l10n-changesets.json
2024-12-11 13:43:08,756 - DEBUG: Fetched https://whattrainisitnow.com/api/release/schedule/?version=135
2024-12-11 13:43:08,894 - DEBUG: Fetched https://whattrainisitnow.com/api/release/schedule/?version=134
2024-12-11 13:43:09,145 - DEBUG: Fetched https://whattrainisitnow.com/api/release/schedule/?version=135
2024-12-11 13:43:09,283 - DEBUG: Fetched https://whattrainisitnow.com/api/release/schedule/?version=134
2024-12-11 13:43:10,816 - DEBUG: Running command: git add ., {'stdin': -3, 'stdout': -1, 'stderr': -1, 'cwd': PosixPath('/tmp/product-details')}
2024-12-11 13:43:10,942 - DEBUG: Running command: git status --short, {'stdin': -3, 'stdout': -1, 'stderr': -1, 'cwd': PosixPath('/tmp/product-details')}
2024-12-11 13:43:11,021 - DEBUG: Running command: git commit -m Updating product details, {'stdin': -3, 'stdout': -1, 'stderr': -1, 'cwd': PosixPath('/tmp/product-details')}
2024-12-11 13:43:11,056 - DEBUG: Running command: git push origin production, {'stdin': -3, 'stdout': -1, 'stderr': -1, 'cwd': PosixPath('/tmp/product-details')}
Username for 'https://github.com': ^Z
[1]+  Stopped                 DEBUG=1 poetry run shipit_rebuild_product_details --database-url="postgresql://shipituser:shipitpassword@db/shipitdb" --channel production
app@afd2542b9914:~$ cd /tmp/product-details
app@afd2542b9914:/tmp/product-details$ git status
On branch production
Your branch is ahead of 'origin/production' by 1 commit.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
app@afd2542b9914:/tmp/product-details$ git show --stat | cat
commit 5b4a4e7fa4c702b4030caa97827691fac5421741
Author: Release Services Robot <release-services+robot@mozilla.com>
Date:   Wed Dec 11 13:43:11 2024 +0000

    Updating product details

 public/1.0/all.json                                |  27 ++
 public/1.0/devedition.json                         |   9 +
 public/1.0/firefox.json                            |  18 +
 .../1.0/firefox_history_development_releases.json  |   3 +-
 public/1.0/firefox_history_stability_releases.json |   3 +-
 public/1.0/firefox_primary_builds.json             | 412 ++++++++++-----------
 public/1.0/firefox_versions.json                   |   8 +-
 public/1.0/l10n/Firefox-133.0.3-build1.json        | 316 ++++++++++++++++
 public/1.0/l10n/Firefox-134.0b8-build1.json        | 316 ++++++++++++++++
 public/1.0/l10n/index.html                         |   2 +
 public/1.0/mobile_android.json                     |   9 +
 public/1.0/mobile_details.json                     |   2 +-
 public/1.0/mobile_versions.json                    |   2 +-
 13 files changed, 913 insertions(+), 214 deletions(-)
```

File: api/src/shipit_api/admin/product_details.py [hunk 1]
```diff
@@ -604,7 +612,12 @@ async def get_primary_builds(
         if Product(release.product) not in products or release.version not in versions:
             continue
         # Make sure to add en-US, it's not listed in the l10n changesets file
-        for l10n in list(releases_l10n.get(release, {}).keys()) + ["en-US"]:
+        locales = ["en-US"]
+        if release in releases_l10n:
+            locales += releases_l10n[release].keys()
+        elif f"1.0/l10n/{release.name}.json" in old_product_details:
```
Comment (bhearsum) [inline api/src/shipit_api/admin/product_details.py:618]: It's annoying that there's two places that we have to consult `old_product_details` :(. In some ideal world, `releases_l10n` would have combined the two data sources already, but I don't think that refactor makes sense in the immediate term. Maybe file for a future improvement, if you agree?

Comment (jcristau) [reply api/src/shipit_api/admin/product_details.py:618]: Filed as #1587.

File: api/src/shipit_api/admin/product_details.py [hunk 2]
```diff
@@ -211,11 +211,19 @@ async def fetch_l10n_data(
     session: aiohttp.ClientSession, release: shipit_api.common.models.Release, raise_on_failure: bool, use_cache: bool = True
 ) -> typing.Tuple[shipit_api.common.models.Release, typing.Optional[ReleaseL10ns]]:
     # Fenix and some thunderbird on the betas don't have l10n in the repository
+    # XXX: for some reason we didn't generate l10n for devedition in old_product_details
```
[the remark this answers is not in the document]
Comment (jcristau) [reply api/src/shipit_api/admin/product_details.py:214]: Filed as #1588.

**Comment by jcristau** [comment]
I can reproduce https://github.com/mozilla-releng/shipit/pull/1586#issuecomment-2541849270, if there's a devedition release without corresponding firefox release, as was the case for ~20 minutes on Friday (Devedition-134.0b10-build1 was marked as shipped at 16:40 UTC, Firefox-134.0b10-build1 at 16:59).
I think we should remove devedition from this sanity_check_firefox_builds function, since there's no devedition_primary_builds file, and firefox_primary_builds only has regular firefox beta.

**Comment by bhearsum** [comment]
> I can reproduce [#1586 (comment)](https://github.com/mozilla-releng/shipit/pull/1586#issuecomment-2541849270), if there's a devedition release without corresponding firefox release, as was the case for ~20 minutes on Friday (Devedition-134.0b10-build1 was marked as shipped at 16:40 UTC, Firefox-134.0b10-build1 at 16:59). I think we should remove devedition from this sanity_check_firefox_builds function, since there's no devedition_primary_builds file, and firefox_primary_builds only has regular firefox beta.

That sounds reasonable to me; the check seems to have outlived whatever value it used to provide.

**Comment by jcristau** [comment]
Looking at this some more... sanity_check_firefox_builds comes from https://github.com/mozilla/release-services/pull/1987.  devedition was then added to get_primary_builds in https://github.com/mozilla/release-services/pull/2111.  So it seems plausible that bedrock needs the FIREFOX_DEVEDITION version to show up in firefox_primary_builds.json.

**Pull request merged**